### ISSUE TITLE
avoid collisions in the population table

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#173]
In the tree sequence tables, nodes refer to rows in the population table; and on output SLiM will write out rows in the population table up to the largest-recorded population ID in the node table. Population `pX` is always in row `X` of the population table. SLiM also writes out metadata for only those populations that are currently present in the simulation. This means that what is "population p1" now might not be "population p1" later, but we have no way of knowing this, leading to difficulties in interpretation.

To fix this, we need to at least:
1. record populations in the population table as soon as they are created, rather than at writing-out time.

Then, we have two options. Tentatively, the first one is preferred, but I'm recording the other one too:

2a. Do not allow the user to create a new subpop `pX` if there is already a population `pX` in the population tables.

2b. Disambiguate SLiM name from row number in the table, storing its SLiM name, so that when SLiM creates a new subpop it adds an entry to the population table, and remembers which tskit-ID it has. This would (only?) require adding a `tsk_subpopulation_id_` property to subpops, and changing [this line](https://github.com/MesserLab/SLiM/blob/915888a1f19123fabbbee7911403f55eaf401086/core/slim_sim.cpp#L5321) from
```
tsk_id_t offspringTSKID = tsk_node_table_add_row(&tables_.nodes, flags, time,
           (tsk_id_t)p_new_genome->subpop_->subpopulation_id_,
```
to
```
tsk_id_t offspringTSKID = tsk_node_table_add_row(&tables_.nodes, flags, time,
           (tsk_id_t)p_new_genome->subpop_->tsk_subpopulation_id_,
```

**Comment by bhaller** [comment]
As long as simplify never reorders/compacts entries in the population table, I think I prefer 2a.  Keep a simple 1-to-1 correspondence between subpop ids and pop table indices.  I am planning to commit 2a shortly.  Can I leave 1 to you?

**Comment by bhaller** [comment]
OK, I think I just committed a fix to 2a, so now just 1 needs to be done.  `Population::AddSubpopulation()` and `Population::AddSubpopulationSplit()` are the two places where that would happen.  Since the SLiMSim object controls all the tree-seq stuff, it'd probably be simplest to just have both methods call a new SLiMSim method, like RegisterNewSubpop(), which would make the new population table entry, including its metadata.  Presumably the existing code that creates the population table at save time would then change accordingly; maybe that method can even be removed entirely, if the population table is always up to date!  Let me know if you need anything from my end.

**Comment by petrelharp** [comment]
> As long as simplify never reorders/compacts entries in the population table,

Yep! `simplify` does not remove populations as long as we don't ask it to: https://tskit.dev/tskit/docs/latest/c-api.html#c.tsk_table_collection_simplify and we are indeed not asking it to: https://github.com/MesserLab/SLiM/blob/master/core/slim_sim.cpp#L5112

And, I'm happy to do #1, but won't realistically do it in the next two weeks (which is OK).

**Comment by bhaller** [comment]
OK; I don't think the things I've committed so far destabilize anything, so this can wait.  Thanks.

**Comment by bhaller** [comment]
Hi @petrelharp, just a ping on you for this.  I'm going through and fixing issues, since I got the proposal out the door.

**Comment by petrelharp** [comment]
Option 2a creates problems if a user creates populations out-of-order (e.g. `p4` before `p3`). But I think this is easily fixable by just checking if the metadata for the population is empty (ie it was just put in as a placeholder) before throwing the error.

**Comment by petrelharp** [comment]
(Note: The code in question is at slim_sim.cpp:5443, at the present moment; SLiMSim::SubpopulationIDInUse().)

**Comment by bhaller** [comment]
With (2a) now fixed – once a p3 has been created, no other subpopulation p3 can be created from then on, even if the old p3 no longer exists – I'm not clear on why (1) also needs to be done.  As I posted above, fixing (1) requires some complicated redesign that I'd prefer not to get into, since this stuff is so brittle and complicated.  @petrelharp can you clarify why (1) still needs to happen?  What's an example of a model where something bad/wrong will now happen, and what exactly is the bad/wrong thing that happens?  Can you provide a MRE?

**Comment by bhaller** [comment]
It would be good to either close this issue or, if there is something that actually needs to be fixed, fix it for SLiM 4.  @petrelharp ping?

**Comment by petrelharp** [comment]
> A big problem with (1) is that we get into the same recording-timing issue that we have for individuals. 

Yes, this is a good point. I've thought about this a fair bit, and I think the motivation was just that everything would be nicer if the tree sequence stored information (ie metadata) about everything it refers to (eg old populations). But, as you say, that's not possible anyhow because populations can change their properties. So, as with individuals, we should just be happy with the tree sequence recording the state at the time it's written out. I haven't been able to think of any actual wrong things that happen, since anytime a `population` is recorded somewhere, it's also recorded along with a `time`, so when interpreting the user just needs to remember that it means "population X at time t", which might not be the same as "population X at the end of the simulation".

So, I think I agree about not needing (1) (yay!).

I do think it's worth thinking *one* more time about (2b), however, given what we've seen people need to do to get tree sequences in to a multispecies sim. Currently, users have to actively stick in "empty populations" into the tree sequence for one species somehow, that exactly correspond to the populations used for another species. This is more complicated/confusing than being able to just set the `slim_id` attribute in metadata to non-conflicting things between the simulations (since it makes people explicitly think about the tables and the order things are added).

**Comment by bhaller** [comment]
> > A big problem with (1) is that we get into the same recording-timing issue that we have for individuals.
> 
> Yes, this is a good point. I've thought about this a fair bit, and I think the motivation was just that everything would be nicer if the tree sequence stored information (ie metadata) about everything it refers to (eg old populations). But, as you say, that's not possible anyhow because populations can change their properties. So, as with individuals, we should just be happy with the tree sequence recording the state at the time it's written out. I haven't been able to think of any actual wrong things that happen, since anytime a `population` is recorded somewhere, it's also recorded along with a `time`, so when interpreting the user just needs to remember that it means "population X at time t", which might not be the same as "population X at the end of the simulation".
> 
> So, I think I agree about not needing (1) (yay!).

Yay!

> I do think it's worth thinking _one_ more time about (2b), however, given what we've seen people need to do to get tree sequences in to a multispecies sim. Currently, users have to actively stick in "empty populations" into the tree sequence for one species somehow, that exactly correspond to the populations used for another species. This is more complicated/confusing than being able to just set the `slim_id` attribute in metadata to non-conflicting things between the simulations (since it makes people explicitly think about the tables and the order things are added).

I'm open to 2b; seems easy (famous last words).  But, here's a random idea that occurred to me as I read this: We could add a mapping table optional parameter to `readFromPopulationFile()` that would just let the user say "the subpopulation 0 in this file should be loaded as subpopulation 3" or whatever.  Is that a good idea, or is it better to make people set up the file with metadata in some way, as with 2b?

**Comment by petrelharp** [comment]
> We could add a mapping table optional parameter to readFromPopulationFile() that would just let the user say "the subpopulation 0 in this file should be loaded as subpopulation 3" or whatever. Is that a good idea, or is it better to make people set up the file with metadata in some way, as with 2b?

I think that'd be way better! It'd make all those headaches go away.

**Comment by bhaller** [comment]
OK, I have just fixed this, I think (in the `multispecies` branch).  I've added a `[No$ subpopMap = NULL]` parameter to `readFromPopulationFile()`.  The relevant doc reads:

> This method can also be used to read tree-sequence (`.trees`) files saved by `treeSeqOutput()` or generated by the Python `pyslim` package.  Beginning with SLiM 4, the `subpopMap` parameter may be supplied to remap the subpopulation ids used in the `.trees` file to different SLiM ids.  This parameter must have a value that is a `Dictionary` (or a subclass of `Dictionary`, such as `DataFrame`) with keys that are the _new_ subpopulation ids (as SLiM-style `string` identifiers like `"p7"`), and values that are the _old_ subpopulation ids (as `integer` values like `7`).  For example, if you pass `Dictionary("p3", 1)` for `subpopMap`, that requests that the subpopulation with an id of `1` in the file should be loaded as subpopulation `p3` in SLiM.  This makes it easier to load simulation data that was generated in Python, since that typically uses an id of `0`.  Note that if `subpopMap` is non-`NULL`, providing a remapping dictionary, then _all_ of the subpopulation ids referenced by the tree sequence must be explicitly remapped in that dictionary (although they can be remapped to their original value – their ids need not all change).  The `subpopMap` parameter may not be used with file formats other than tree-sequence files, at the present time; setting up the correct subpopulation ids is typically easier when working with those other formats.

Hopefully this will be what is needed.  It would be great to get some use of this from folks like @clwgg and @bobweek who have seemed to be interested in related stuff.  @petrelharp I take back what I said about diff review; the diffs on this are pretty big and hairy, and I'm less worried about it than I was because really the new code doesn't affect anything at all unless a remapping dictionary is supplied; the risk of this breaking any existing code is close to nil, I think.  A better use of time would be to get the python-based test suite running again, and then perhaps to add some new multispecies tests that exercise this and other changes that I've recently made to the tree-seq stuff.  :->

Thanks all.  Nice to close such an old issue, finally!

**Comment by bhaller** [comment]
Oh, and: in case it is useful to see an example, here's a script that reads in .trees files for each of two species in a model:

```
species all initialize() {
}
function (void)sharedInit(i$ mutTypeID, i$ geTypeID) {
	initializeTreeSeq(runCrosschecks=T);
	initializeMutationRate(1e-8);
	initializeMutationType(mutTypeID, 0.5, "f", 0.0);
	initializeGenomicElementType(geTypeID, mutTypeID, 1.0);
	initializeGenomicElement(geTypeID, 0, 1e6-1);
	initializeRecombinationRate(1e-8);
}
species sp1 initialize() {
	// sp1 uses m1 and g1; its .trees file needs to reference those ids
	sharedInit(1, 1);
}
species sp2 initialize() {
	// sp1 uses m2 and g2; its .trees file needs to reference those ids
	sharedInit(2, 2);
}

ticks all 1 late() {
	sp1.readFromPopulationFile("/Users/bhaller/Desktop/pop_p1_1.trees",
		subpopMap=Dictionary("p3", 1));
	sp2.readFromPopulationFile("/Users/bhaller/Desktop/pop_p1_2.trees",
		subpopMap=Dictionary("p7", 1));
}

ticks all 10000 late() {
}
```

This script remaps id 1 to p3 for species sp1, and id 1 to p7 for species sp2.  As the comments note, the mutation types and genomic element types used in the files do still need to match what is defined in the SLiM model.  I made the .trees files loaded here using two other SLiM models that had the appropriate genetic setup.  If there is a need to provide the ability to remap those as well, feel free to file a new issue; that would not be particularly difficult (much easier than this was).  But my impression is that setting those up to use the right ids is not hard on the Python side, so I'd prefer not to clutter the API up with multiple remapping dictionaries if we don't have to.

**Comment by petrelharp** [comment]
Looks perfect (from the docs =). Very minor nitpick: maybe remove the "(or a subclass of Dictionary...)" note in the docstring?

**Comment by bobweek** [comment]
> Hopefully this will be what is needed. It would be great to get some use of this from folks like @clwgg and @bobweek who have seemed to be interested in related stuff.

I'm still getting an error: `property name must be unique across all subpopulations; pop_0 is already in use, or was previously used.` attached are (hopefully) mwe's to recreate this error.

[minimal-burn-in.py.txt](https://github.com/MesserLab/SLiM/files/8694206/minimal-burn-in.py.txt)
[mwe.slim.txt](https://github.com/MesserLab/SLiM/files/8694207/mwe.slim.txt)

**Comment by bhaller** [comment]
Both subpopulations are named `pop_0`, which causes a conflict.  If they were named `p0`, that name would get remapped by the mapping table, since that name would follow SLiM's defined naming rules.  The name `pop_0` does not, and so a conflict results.  Where does that name come from?  Anyhow, if you set it to be `p0` then everything ought to work fine; but perhaps this is a place where additional standardization could smooth things out somehow.  Anyway, I don't _think_ this is a SLiM bug at this point.

**Comment by bhaller** [comment]
I suppose SLiM could remap `pop_0` as well, potentially.  But then will we need to remap `population_0` and `subpopulation_0` and `pop0` and so on and so forth?  I guess I'd like to know where the name `pop_0` is coming from and whether it represents a standard defined by some package that SLiM ought to respect, or whether it is just an arbitrary name?  @petrelharp are you familiar with this?

**Comment by bhaller** [comment]
Reopening for now.

**Comment by petrelharp** [comment]
> > Looks perfect (from the docs =). Very minor nitpick: maybe remove the "(or a subclass of Dictionary...)" note in the docstring?
> 
> Because that is implied by object-oriented programming in general? Or for some other reason?

Yes, it's implied, and I could imagine you wanting to change things in the future so that it *didn't*work for any subclass of Dictionary.

**Comment by petrelharp** [comment]
Well, this is a tricky one. We've got to stick in something for the `name` when we annotate. So, Bob could rename the populations so that the names don't conflict, but then that seems like it defeats the purpose of remapping. The remapping dict could have values that are population names (instead of IDs), but we'd still have to reassign new names somehow. 

And, confusingly, changing the first population's name after reading it in doesn't fix the error:
```
ticks all 1 late() {
	sp1.readFromPopulationFile("hinit.trees",
		subpopMap=Dictionary("p1", 0));
	p1.name = "foo";
	sp2.readFromPopulationFile("pinit.trees",
		subpopMap=Dictionary("p2", 0));
```

**Comment by petrelharp** [comment]
If renaming the populations worked to avoid conflicts, I'd say that's a good and easy workaround - and, I don't see why it shouldn't work?

**Comment by bhaller** [comment]
I'm not sure I understand all the issues at play here; but is it really necessary to use "pop_0" instead of following the SLiM convention of calling it "p0"?  That is the root of the problem, as far as I can see.  If you give both populations names that are the same, and thus conflict, like naming them both "foo", then you get a conflict.  SLiM doesn't know that there is no conflict unless you follow its naming convention – "p0", not "pop_0".  Of course these rules can be changed, but it is not clear to me exactly what they should be changed to, and I'm wondering whether it is not possible to just use "p0" instead to solve the problem...?

**Comment by bhaller** [comment]
> ...And, confusingly, changing the first population's name after reading it in doesn't fix the error...

Yes; I believe this is the policy we settled upon when we added the name metadata for subpops recently, no?  No name or id can be reused, even if it is no longer in use now, to avoid conflicts between the names/ids or subpops at different times.

**Comment by bhaller** [comment]
Sorry if I'm missing the point here.  Exhausted, going to bed.

**Comment by bhaller** [comment]
> Ah, I see. Well, getting things named `p0` instead of `pop_0` would fix the issue, but that's not totally straightforward, since msprime's defaults are to use `pop_0`, and I don't think `pyslim.annotate_defaults` should change population names. I suppose it could change the names to `pX` _if_ they are `pop_X`, but that's a bit ugly. Or, I guess, SLiM could deal with `pop_X` in the same way as `pX` when reading in, but that's maybe worse. I need to thiink about this more.

If `pop_0` is an msprime default that is standard (i.e., not a moving target) then I'm OK with having SLiM handle `pop_0` the same way as `p0`, I think (but you should ponder that since you understand the msprime side far better than I).  I just don't want to get into the game of trying to resolve conflicts with arbitrary/non-standard names.  If `po_0` is standard then this is just interoperability, no problem with that.  Shall I proceed with that fix?

**Comment by bhaller** [comment]
After discussion, the consensus was yes, proceed with remapping `pop_0` to `pX` according to the mapping dictionary.

**Comment by bhaller** [comment]
Just fixed in `multispecies`, closing.
